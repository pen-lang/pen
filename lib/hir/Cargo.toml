[package]
name = "hir"
version = "0.1.0"
authors = ["Yota Toyama <raviqqe@gmail.com>"]
edition = "2021"

[dependencies]
fnv = "1"
position = { path = "../position" }
serde = { version = "1", features = ["derive", "rc"] }

[dev-dependencies]
once_cell = "1"
pretty_assertions = "1"
