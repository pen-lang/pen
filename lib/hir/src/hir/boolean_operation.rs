use super::expression::Expression;
use position::Position;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum BooleanOperator {
    And,
    Or,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BooleanOperation {
    operator: BooleanOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    position: Position,
}

impl BooleanOperation {
    pub fn new(
        operator: BooleanOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            operator,
            lhs: lhs.into().into(),
            rhs: rhs.into().into(),
            position,
        }
    }

    pub fn operator(&self) -> BooleanOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
