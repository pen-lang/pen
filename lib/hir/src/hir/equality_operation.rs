use super::expression::Expression;
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum EqualityOperator {
    Equal,
    NotEqual,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EqualityOperation {
    type_: Option<Type>,
    operator: EqualityOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    position: Position,
}

impl EqualityOperation {
    pub fn new(
        type_: Option<Type>,
        operator: EqualityOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            type_,
            operator,
            lhs: Arc::new(lhs.into()),
            rhs: Arc::new(rhs.into()),
            position,
        }
    }

    pub fn type_(&self) -> Option<&Type> {
        self.type_.as_ref()
    }

    pub fn operator(&self) -> EqualityOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
