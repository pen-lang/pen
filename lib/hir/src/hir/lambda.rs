use super::{Argument, Expression};
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct Lambda {
    arguments: Vec<Argument>,
    result_type: Type,
    body: Arc<Expression>,
    position: Position,
}

impl Lambda {
    pub fn new(
        arguments: Vec<Argument>,
        result_type: impl Into<Type>,
        body: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            arguments,
            result_type: result_type.into(),
            body: body.into().into(),
            position,
        }
    }

    pub fn arguments(&self) -> &[Argument] {
        &self.arguments
    }

    pub fn result_type(&self) -> &Type {
        &self.result_type
    }

    pub fn body(&self) -> &Expression {
        &self.body
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
