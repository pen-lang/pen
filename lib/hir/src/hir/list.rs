use super::list_element::ListElement;
use crate::types::Type;
use position::Position;

#[derive(Clone, Debug, PartialEq)]
pub struct List {
    type_: Type,
    elements: Vec<ListElement>,
    position: Position,
}

impl List {
    pub fn new(type_: impl Into<Type>, elements: Vec<ListElement>, position: Position) -> Self {
        Self {
            type_: type_.into(),
            elements,
            position,
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn elements(&self) -> &[ListElement] {
        &self.elements
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
