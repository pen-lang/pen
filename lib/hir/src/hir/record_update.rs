use super::{expression::Expression, RecordElement};
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct RecordUpdate {
    type_: Type,
    record: Arc<Expression>,
    elements: Vec<RecordElement>,
    position: Position,
}

impl RecordUpdate {
    pub fn new(
        type_: impl Into<Type>,
        record: impl Into<Expression>,
        elements: Vec<RecordElement>,
        position: Position,
    ) -> Self {
        Self {
            type_: type_.into(),
            record: Arc::new(record.into()),
            elements,
            position,
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn record(&self) -> &Expression {
        &self.record
    }

    pub fn elements(&self) -> &[RecordElement] {
        &self.elements
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
