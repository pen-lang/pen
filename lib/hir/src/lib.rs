pub mod hir;
mod test;
pub mod types;
