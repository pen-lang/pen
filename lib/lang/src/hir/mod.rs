pub mod analysis;
mod argument;
mod arithmetic_operation;
mod boolean;
mod boolean_operation;
mod call;
mod declaration;
mod definition;
mod else_branch;
mod equality_operation;
mod expression;
mod if_;
mod if_list;
mod if_type;
mod if_type_branch;
mod lambda;
mod let_;
mod list;
mod list_element;
mod module;
mod none;
mod not_operation;
mod number;
mod operation;
mod order_operation;
mod record_construction;
mod record_deconstruction;
mod record_element;
mod record_update;
mod string;
mod try_operation;
mod type_alias;
mod type_coercion;
mod type_definition;
mod variable;

pub use argument::*;
pub use arithmetic_operation::*;
pub use boolean::*;
pub use boolean_operation::*;
pub use call::*;
pub use declaration::*;
pub use definition::*;
pub use else_branch::*;
pub use equality_operation::*;
pub use expression::*;
pub use if_::*;
pub use if_list::*;
pub use if_type::*;
pub use if_type_branch::*;
pub use lambda::*;
pub use let_::*;
pub use list::*;
pub use list_element::*;
pub use module::*;
pub use none::*;
pub use not_operation::*;
pub use number::*;
pub use operation::*;
pub use order_operation::*;
pub use record_construction::*;
pub use record_deconstruction::*;
pub use record_element::*;
pub use record_update::*;
pub use string::*;
pub use try_operation::*;
pub use type_alias::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use variable::*;
