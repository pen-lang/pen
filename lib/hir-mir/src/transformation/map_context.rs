pub mod expression;
pub mod module;

use crate::CompileError;
use fnv::FnvHashMap;
use hir::{
    analysis::type_id_calculator,
    types::{self, Type},
};

pub(self) fn context_function_name(
    type_: &types::Map,
    types: &FnvHashMap<String, Type>,
) -> Result<String, CompileError> {
    Ok(format!(
        "hir:map:context:{}",
        type_id_calculator::calculate(&type_.clone().into(), types)?
    ))
}
