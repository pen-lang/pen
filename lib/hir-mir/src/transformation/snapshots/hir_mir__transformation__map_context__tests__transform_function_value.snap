---
source: lib/hir-mir/src/transformation/map_context.rs
expression: "transform_module(&context,\n    &Module::empty().set_function_definitions(vec![FunctionDefinition ::\n                fake(\"f\", Lambda ::\n                new(vec! [], collection_type ::\n                transform_map_context(& context, & Position ::\n                fake()).unwrap(), Map ::\n                new(types :: None :: new(Position :: fake()), types ::\n                Function ::\n                new(vec! [], types :: None :: new(Position :: fake()),\n                Position :: fake()), vec! [], Position :: fake()), Position ::\n                fake()), false)]))"
---
Ok(
    Module {
        type_definitions: [],
        type_aliases: [],
        foreign_declarations: [],
        function_declarations: [],
        function_definitions: [
            FunctionDefinition {
                name: "f",
                original_name: "",
                lambda: Lambda {
                    arguments: [],
                    result_type: Reference(
                        Reference {
                            name: "mapContext",
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    body: Map(
                        Map {
                            key_type: None(
                                None {
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            value_type: Function(
                                Function {
                                    arguments: [],
                                    result: None(
                                        None {
                                            position: Position {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        },
                                    ),
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            elements: [],
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 1,
                        column_number: 1,
                        line: "",
                    },
                },
                foreign_definition_configuration: None,
                public: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
            FunctionDefinition {
                name: "hir:map:context:8bbb62926defb736",
                original_name: "hir:map:context:8bbb62926defb736",
                lambda: Lambda {
                    arguments: [],
                    result_type: Reference(
                        Reference {
                            name: "mapContext",
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    body: Call(
                        Call {
                            function_type: Some(
                                Function(
                                    Function {
                                        arguments: [
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Boolean(
                                                        Boolean {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Number(
                                                        Number {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Boolean(
                                                        Boolean {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Number(
                                                        Number {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                        ],
                                        result: Reference(
                                            Reference {
                                                name: "mapContext",
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                            ),
                            function: Variable(
                                Variable {
                                    name: "newMapContext",
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            arguments: [
                                Lambda(
                                    Lambda {
                                        arguments: [
                                            Argument {
                                                name: "$lhs",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                            Argument {
                                                name: "$rhs",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                        ],
                                        result_type: Boolean(
                                            Boolean {
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        body: IfType(
                                            IfType {
                                                name: "$lhs",
                                                argument: Variable(
                                                    Variable {
                                                        name: "$lhs",
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                branches: [
                                                    IfTypeBranch {
                                                        type_: None(
                                                            None {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        expression: IfType(
                                                            IfType {
                                                                name: "$rhs",
                                                                argument: Variable(
                                                                    Variable {
                                                                        name: "$rhs",
                                                                        position: Position {
                                                                            path: "",
                                                                            line_number: 1,
                                                                            column_number: 1,
                                                                            line: "",
                                                                        },
                                                                    },
                                                                ),
                                                                branches: [
                                                                    IfTypeBranch {
                                                                        type_: None(
                                                                            None {
                                                                                position: Position {
                                                                                    path: "",
                                                                                    line_number: 1,
                                                                                    column_number: 1,
                                                                                    line: "",
                                                                                },
                                                                            },
                                                                        ),
                                                                        expression: Boolean(
                                                                            Boolean {
                                                                                value: true,
                                                                                position: Position {
                                                                                    path: "",
                                                                                    line_number: 1,
                                                                                    column_number: 1,
                                                                                    line: "",
                                                                                },
                                                                            },
                                                                        ),
                                                                    },
                                                                ],
                                                                else_: None,
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    },
                                                ],
                                                else_: None,
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                                Lambda(
                                    Lambda {
                                        arguments: [
                                            Argument {
                                                name: "$x",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                        ],
                                        result_type: Number(
                                            Number {
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        body: IfType(
                                            IfType {
                                                name: "$x",
                                                argument: Variable(
                                                    Variable {
                                                        name: "$x",
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                branches: [
                                                    IfTypeBranch {
                                                        type_: None(
                                                            None {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        expression: Number(
                                                            Number {
                                                                value: 0.0,
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    },
                                                ],
                                                else_: None,
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                                Lambda(
                                    Lambda {
                                        arguments: [
                                            Argument {
                                                name: "",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                            Argument {
                                                name: "",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                        ],
                                        result_type: Boolean(
                                            Boolean {
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        body: Boolean(
                                            Boolean {
                                                value: false,
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                                Lambda(
                                    Lambda {
                                        arguments: [
                                            Argument {
                                                name: "",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                        ],
                                        result_type: Number(
                                            Number {
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        body: Number(
                                            Number {
                                                value: 0.0,
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                            ],
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 1,
                        column_number: 1,
                        line: "",
                    },
                },
                foreign_definition_configuration: None,
                public: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
        ],
        position: Position {
            path: "",
            line_number: 1,
            column_number: 1,
            line: "",
        },
    },
)
